kf/bpf.go

```diff
@@ -1191,25 +1118,79 @@ func (b *BPF) UnloadProgram(ifaceName, direction string) error {
 	}
 
 	// remove pinned map file
-	if err := b.RemoveMapFile(); err != nil {
+	if err := b.RemoveMapFiles(ifaceName); err != nil {
 		log.Error().Msgf("failed to remove map file for program %s => %s", ifaceName, b.Program.Name)
 	}
 
 	return nil
 }
 
-func (b *BPF) RemoveMapFile() error {
-	if err := os.RemoveAll(b.MapNamePath); os.IsNotExist(err) {
-		log.Info().Msgf("RemoveMapFile: map file removed successfully - %s ", b.MapNamePath)
-		return nil
+// RemoveMapFiles - removes all the pinned map files
+func (b *BPF) RemoveMapFiles(ifaceName string) error {
+	// For tc and xdp programs pinning the maps only if there is a user program needs to access.
+	if len(b.Program.CmdStart) > 0 || len(b.Program.CmdUpdate) > 0 || len(b.Program.CmdStop) > 0 || len(b.Program.CmdStatus) > 0 {
+		for k, v := range b.ProgMapCollection.Maps {
+			var mapFilename string
+			if b.Program.ProgType == models.TCType {
+				mapFilename = filepath.Join(b.hostConfig.BpfMapDefaultPath, models.TCMapPinPath, ifaceName, k)
+			} else {
+				mapFilename = filepath.Join(b.hostConfig.BpfMapDefaultPath, ifaceName, k)
+			}
+			if err := v.Unpin(); err != nil {
+				return fmt.Errorf("eBPF program %s prog type %s ifacename %s map %s:failed to pin the map err - %#v",
+					b.Program.Name, b.Program.ProgType, ifaceName, mapFilename, err)
+			}
+			if strings.Contains(mapFilename, "..") {
+				return fmt.Errorf("eBPF program %s prog type %s ifacename %s map %s:contains relative path",
+					b.Program.Name, b.Program.ProgType, ifaceName, mapFilename)
+			}
+			if err := os.RemoveAll(mapFilename); os.IsNotExist(err) {
+				log.Info().Msgf("RemoveMapFile: tc program %s map file removed successfully - %s err - %#v", b.Program.Name, mapFilename, err)
+			}
+		}
+	} else { // no user program accessing pin
+		ss := strings.Split(b.Program.MapName, "/")
+		ebpfMap := b.ProgMapCollection.Maps[ss[len(ss)-1]]
+		if err := ebpfMap.Unpin(); err != nil {
+			log.Warn().Msgf("eBPF program %s map %s:failed to unpin the map error %#v", b.Program.Name, b.Program.MapName, err)
+		}
+		if err := os.RemoveAll(b.MapNamePath); os.IsNotExist(err) {
+			log.Info().Msgf("RemoveMapFile: xdp program %s map file removed successfully - %s err - %#v", b.Program.Name, b.MapNamePath, err)
```

Comment:
```suggestion
			log.Info().Msgf("RemoveMapFiles: xdp program %s map file removed successfully - %s err - %#v", b.Program.Name, b.MapNamePath, err)
```
if these were meant to have the function name, then correct the function name